File: sol/2563/2563.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        n = len(nums)
        res = 0
        l = n
        r = n - 1
        for i, x in enumerate(nums):
            l = max(i + 1, l)
            while l > i + 1 and x + nums[l - 1] >= lower:
                l -= 1
            while r >= l and x + nums[r] > upper:
                r -= 1
            if r <= i:
                break
            res += r - l + 1
        return res

File: sol/2563/2563.cpp
struct Solution {
    using ll = long long;
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        ranges::sort(nums);
        int n = nums.size();
        ll res = 0;
        int l = n, r = n - 1;
        for (int i = 0; i < n; ++i){
            l = max(l, i + 1);
            while (l > i + 1 && nums[i] + nums[l - 1] >= lower){
                --l;
            }
            while (r >= l && nums[i] + nums[r] > upper){
                --r;
            }
            if (r <= i){
                break;
            }
            res += r - l + 1;
        }
        return res;
    }
};
